### PR TITLE
[Feature] ✨ adding support for OpenTelemetry (OTEL)

### PR DESCRIPTION
**Pull request by mfreeman451**
# Description

Adding support for OpenTelemetry (OTLP) Exporters.

Instrumented:

- [x] mage-ai-server (tornado, sqlalchemy)
- [x] mage-app-server (nextjs experimental instrumentation)

Screenshot 2023-11-25 at 4 23 27 PM

Screenshot 2023-11-25 at 6 12 37 PM

This pull request introduces significant enhancements to the monitoring capabilities of our `mage_ai` application by integrating OpenTelemetry. The changes span across both the Python backend and the Next.js/Node.js frontend, ensuring comprehensive telemetry data collection.

### Backend (Python):
- **Added OpenTelemetry Libraries**: Integrated OpenTelemetry Python libraries to instrument the application for telemetry data collection.
  - `opentelemetry-api`: Provides the API for tracing and metrics.
  - `opentelemetry-sdk`: The core SDK for OpenTelemetry in Python.
  - `opentelemetry-instrumentation`: General instrumentation library.
  - `opentelemetry-instrumentation-sqlalchemy`: Specific for instrumenting SQLAlchemy operations.
  - `opentelemetry-instrumentation-tornado`: For monitoring Tornado-based HTTP server activities.
- **Environment Variables**: Added new environment variables (`OTEL_EXPORTER_OTLP_HTTP_ENDPOINT` and `OTEL_EXPORTER_OTLP_ENDPOINT`) for configuring the OpenTelemetry Collector endpoints.
- **Database and HTTP Server Instrumentation**: Configured instrumentation for both SQLAlchemy database operations and Tornado HTTP server activities to gather traces and metrics.

### Frontend (Next.js/Node.js):
  - **Added OpenTelemetry Libraries for Node.js**: Integrated OpenTelemetry Node.js libraries for monitoring frontend operations.
  - `@opentelemetry/api`: Core OpenTelemetry API for Node.js.
  - `@opentelemetry/sdk-node`: SDK implementation for Node.js.
  - `@opentelemetry/auto-instrumentations-node`: Automatically instruments Node.js applications.
  - **Configuration for Telemetry Data Export**: Set up the frontend to export telemetry data to the specified OpenTelemetry Collector, in line with the backend configuration.
  - **nextjs upgrade to 14.0.3**: Enables experimental instrumentation feature

## Libraries Added

- **Python:**
  - `opentelemetry-api`
  - `opentelemetry-sdk`
  - `opentelemetry-instrumentation`
  - `opentelemetry-instrumentation-sqlalchemy`
  - `opentelemetry-instrumentation-tornado`

- **Next.js/Node.js:**
  - `@opentelemetry/api`
  - `@opentelemetry/sdk-node`
  - `@opentelemetry/auto-instrumentations-node`

## Impact

- **Enhanced Monitoring**: The integration of OpenTelemetry provides enhanced visibility into both backend and frontend operations, enabling detailed monitoring and troubleshooting.
- **Scalable Observability**: The setup paves the way for scalable observability practices within the application, making it easier to monitor, debug, and optimize performance.
- **Unified Telemetry Data**: With both frontend and backend telemetry data being collected, we have a unified view of the application's performance metrics and traces.

# How Has This Been Tested?

- **Backend Testing**: Conducted tests to ensure SQLAlchemy and Tornado instrumentation effectively capture relevant telemetry data.
- **Frontend Testing**: Verified that the Next.js application correctly exports telemetry data to the configured OpenTelemetry Collector.
- **End-to-End Monitoring**: Tested the entire flow of telemetry data from the application to the OpenTelemetry Collector and onto our observability platform.

# Checklist
- [x] The PR is tagged with proper labels (bug, enhancement, feature, documentation)
- [x] I have performed a self-review of my own code
- [ ] I have added unit tests that prove my fix is effective or that my feature works
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] If new documentation has been added, relative paths have been added to the appropriate section of `docs/mint.json`

cc:

### CODE REVIEW

File: dev.Dockerfile [hunk 1]
```diff
@@ -1,6 +1,10 @@
 FROM python:3.10-bookworm
 LABEL description="Mage data management platform"
 ARG PIP=pip3
+ARG OTEL_EXPORTER_OTLP_ENDPOINT
+ARG OTEL_EXPORTER_OTLP_HTTP_ENDPOINT
+ENV OTEL_EXPORTER_OTLP_HTTP_ENDPOINT=${OTEL_EXPORTER_OTLP_HTTP_ENDPOINT}
```
Comment (wangxiaoyou1993) [inline dev.Dockerfile:6]: You don't need to add these environment variables in the dockerfile.

File: docker-compose.yml [hunk 2]
```diff
@@ -4,7 +4,12 @@ x-server_settings: &server_settings
   build:
     context: .
     dockerfile: ./dev.Dockerfile
+    args:
+      - OTEL_EXPORTER_OTLP_ENDPOINT=${OTEL_EXPORTER_OTLP_ENDPOINT}
```
Comment (wangxiaoyou1993) [inline docker-compose.yml:8]: If you already set the variables in environment variables below, you don't need to pass them as args to the Dockerfile.

File: mage_ai/server/server.py [hunk 3]
```diff
@@ -333,14 +334,50 @@ def make_app(template_dir: str = None, update_routes: bool = False):
         (r'/version-control', MainPageHandler),
     ]
 
+    if ENABLE_PROMETHEUS or OTEL_EXPORTER_OTLP_ENDPOINT:
+        from opentelemetry.instrumentation.tornado import TornadoInstrumentor
+        TornadoInstrumentor().instrument()
+        print('OpenTelemetry instrumentation enabled.')
```
Comment (wangxiaoyou1993) [inline mage_ai/server/server.py:340]: nit: you can use logger.info

File: mage_ai/server/server.py [hunk 4]
```diff
@@ -333,14 +334,50 @@ def make_app(template_dir: str = None, update_routes: bool = False):
...
+
+    if OTEL_EXPORTER_OTLP_ENDPOINT:
+        print(f'OTEL_EXPORTER_OTLP_ENDPOINT: {OTEL_EXPORTER_OTLP_ENDPOINT}')
```
Comment (wangxiaoyou1993) [inline mage_ai/server/server.py:343]: nit: same here, you can use logger.info

File: mage_ai/orchestration/db/__init__.py [hunk 5]
```diff
@@ -3,6 +3,7 @@
 from urllib.parse import parse_qs, quote_plus, urlparse
 
 import sqlalchemy
+from opentelemetry.instrumentation.sqlalchemy import SQLAlchemyInstrumentor
```
Comment (wangxiaoyou1993) [inline mage_ai/orchestration/db/__init__.py:6]: Let's only import and use it when OTEL feature is enabled. Then opentelemetry libraries are not installed in the base Mage package.

**Comment by johnson-mage** [comment]
When trying to build a fresh docker image off this branch, I ran into an error that caused my build to fail (the docker build didn't have any issues for me on the `master` branch):
image
image

**Comment by johnson-mage** [comment]
> When trying to build a fresh docker image off this branch, I ran into an error that caused my build to fail (the docker build didn't have any issues for me on the `master` branch): image image

@mfreeman451 Can you please rebase the `master` branch when you get a chance? This build issue should be resolved now.

**Comment by mfreeman451** [comment]
Will do, thank you!

File: scripts/dev.sh [hunk 6]
```diff
@@ -153,6 +171,9 @@ export REQUIRE_USER_AUTHENTICATION=$REQUIRE_USER_AUTHENTICATION
 export REQUIRE_USER_PERMISSIONS=$REQUIRE_USER_PERMISSIONS
 export DEBUG=$DEBUG
 
+
+env
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply scripts/dev.sh]: removed

File: dev.Dockerfile (hunk 1 above)
Comment (mfreeman451) [reply dev.Dockerfile]: removed

File: mage_ai/server/server.py (hunk 3 above)
Comment (mfreeman451) [reply mage_ai/server/server.py]: updated

**Pull request closed without merging**